pkg/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,121 @@
+package selfupdate
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/google/go-github/github"
+	"github.com/minio/selfupdate"
+)
+
+type SelfUpdater struct {
+	assetId     int64
+	githubOwner string
+	githubRepo  string
+	assetUrl    string
+	Version     semver.Version
+}
+
+func NewSelfUpdater(ctx context.Context, githubOwner, githubRepo string) (*SelfUpdater, error) {
+	var err error
+	var githubRelease *github.RepositoryRelease
+
+	selfUpdater := new(SelfUpdater)
+	selfUpdater.githubOwner = githubOwner
+	selfUpdater.githubRepo = githubRepo
+
+	client := github.NewClient(nil)
+	if githubRelease, _, err = client.Repositories.GetLatestRelease(ctx, githubOwner, githubRepo); err != nil {
+		return nil, err
+	}
+	if err = selfUpdater.fetchVersion(githubRelease); err != nil {
+		return nil, err
+	}
+	if err = selfUpdater.fetchAsset(ctx, client, githubRelease); err != nil {
+		return nil, err
+	}
+	return selfUpdater, err
+}
+
+func (selfUpdater *SelfUpdater) fetchVersion(githubRelease *github.RepositoryRelease) error {
+	var err error
+
+	selfUpdater.Version, err = semver.ParseTolerant(githubRelease.GetTagName())
```

Comment:
what happens to version if err != nil ? do we still want the assignment? might be risky